### PR TITLE
semgrep{,-core}: 1.15.0 -> 1.27.0

### LINKED ISSUE #235417: Update request: semgrep 1.15 → 1.24
- Package name: semgrep
- Latest released version:1.24

- Current version on the unstable channel: 1.15
- Current version on the stable/release channel: 1.15

- [ x] Checked the [nixpkgs pull requests](https://github.com/NixOS/nixpkgs/pulls)

**Notify maintainers**

@06kellyjac 
@ambroisie 

-----

Note for maintainers: Please tag this issue in your PR.

### PR DESCRIPTION
**Pull request by 06kellyjac**
###### Description of changes

Can no longer fetch prebuilt semgrep-core from github but can get it from the python wheel
We're still able to build the rest of semgrep from source

---

related: https://github.com/returntocorp/semgrep/issues/7570

resolves: #235417

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [X] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes (or backporting 23.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by 06kellyjac** [comment]
TODO:

Fixup semgrep-core

```
λ ./result/bin/semgrep scan --metrics off --lang yaml --config 'r/generic.unicode.security.bidi.contains-bidirectional-characters'
               
               
┌─────────────┐
│ Scan Status │
└─────────────┘
  Scanning 35542 files tracked by git with 1 Code rule:
  Scanning 1237 files.
  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━   0% -:--:--                                                                                                                        
[Errno 13] Permission denied: '/nix/store/cwkvhrqk62b6qs2lg02yy0n4h5pld9g7-semgrep-1.24.1/lib/python3.10/site-packages/semgrep/bin/semgrep-core'
Traceback (most recent call last):
  File "/nix/store/cwkvhrqk62b6qs2lg02yy0n4h5pld9g7-semgrep-1.24.1/lib/python3.10/site-packages/semgrep/commands/wrapper.py", line 35, in wrapper
    func(*args, **kwargs)
  File "/nix/store/cwkvhrqk62b6qs2lg02yy0n4h5pld9g7-semgrep-1.24.1/lib/python3.10/site-packages/semgrep/commands/scan.py", line 907, in scan
# ...
```

I thought I'd get the ball rolling on the PR so someone can help me with the fetchpypi bit.

**Comment by 06kellyjac** [comment]
Fixed, it came with a static semgrep-core but it wasn't marked executable for some reason...

```
λ ./result/bin/semgrep scan --metrics off --lang yaml --config 'r/generic.unicode.security.bidi.contains-bidirectional-characters'
               
               
┌─────────────┐
│ Scan Status │
└─────────────┘
  Scanning 35542 files tracked by git with 1 Code rule:
  Scanning 1237 files.
  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00                                                                                                                        
                
                
┌──────────────┐
│ Scan Summary │
└──────────────┘
Some files were skipped or only partially analyzed.
  Scan was limited to files tracked by git.
  Scan skipped: 2 files larger than 1.0 MB, 89 files matching .semgrepignore patterns
  For a full list of skipped files, run semgrep with the --verbose flag.

(need more rules? `semgrep login` for additional free Semgrep Registry rules)

Ran 1 rule on 1237 files: 0 findings.
If Semgrep missed a finding, please send us feedback to let us know!
  $ semgrep shouldafound --help
```

**Comment by 06kellyjac** [comment]
@mweinelt since you do a bit of python would you mind reviewing this & maybe helping with fetching the wheel properly :upside_down_face:

File: pkgs/tools/security/semgrep/default.nix [hunk 1]
```diff
@@ -1,41 +1,28 @@
...
-        ln -s ${submodule}/ ${path}
-      ''
-    )
-    common.submodules)) + ''
-    cd cli
-  '';
+  src = fetchurl {
+    url = "https://files.pythonhosted.org/packages/33/c5/c236d76a2b7501355e4957440b15aeb6c375b58bd1adb6ed5af16585cad1/semgrep-1.24.1-cp37.cp38.cp39.py37.py38.py39-none-any.whl";
+    hash = "sha256-qpBchqKJBJcg3uHLuTnF3y9yzAb5yKI2Vm927WgzQQI=";
+  };
```
[the remark this answers is not in the document]
Comment (06kellyjac) [reply pkgs/tools/security/semgrep/default.nix:23]: Looks like it is the same as `python` here

https://github.com/BlackArch/blackarch/blob/master/packages/semgrep/PKGBUILD#L23

**Comment by 06kellyjac** [comment]
Result of `nixpkgs-review pr 235466` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>semgrep</li>
    <li>semgrep-core</li>
    <li>semgrep.dist</li>
  </ul>
</details>

**Review by ambroisie — APPROVED** [review]

**Comment by 06kellyjac** [comment]
@ambroisie I realised even though the wheel is "none-any" it isn't actually going to be a cross platform binary & I didn't remove the darwin platforms.

I've now changed the updater and fetching for semgrep-core to actually get the darwin copies of the wheels

**Comment by reckenrode** [comment]
Result of `nixpkgs-review pr 235466` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>semgrep</li>
    <li>semgrep-core</li>
    <li>semgrep.dist</li>
  </ul>
</details>

**Comment by reckenrode** [comment]
> Thanks @reckenrode! Did you try run `semgrep`? e.g. `./result/bin/semgrep scan --metrics off --lang yaml --config 'r/generic.unicode.security.bidi.contains-bidirectional-characters'`

This is what I get running that command.

```console
$ ./results/semgrep/bin/semgrep  scan --metrics off --lang yaml --config 'r/generic.unicode.security.bidi.contains-bidirectional-characters'


┌─────────────┐
│ Scan Status │
└─────────────┘
  Scanning 7 files tracked by git with 1 Code rule:
  Scanning 1 file.
  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00


┌──────────────┐
│ Scan Summary │
└──────────────┘

(need more rules? `semgrep login` for additional free Semgrep Registry rules)

Ran 1 rule on 1 file: 0 findings.
If Semgrep missed a finding, please send us feedback to let us know!
  $ semgrep shouldafound --help
```

**Comment by reckenrode** [comment]
Result of `nixpkgs-review pr 235466` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>semgrep</li>
    <li>semgrep-core</li>
    <li>semgrep.dist</li>
  </ul>
</details>

**Comment by reckenrode** [comment]
```
$ ./results/semgrep/bin/semgrep  scan --metrics off --lang yaml --config 'r/generic.unicode.security.bidi.contains-bidirectional-characters'


┌─────────────┐
│ Scan Status │
└─────────────┘
  Scanning 35634 files tracked by git with 1 Code rule:
  Scanning 1220 files.
  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00


┌──────────────┐
│ Scan Summary │
└──────────────┘
Some files were skipped or only partially analyzed.
  Scan skipped: 2 files larger than 1.0 MB, 89 files matching .semgrepignore patterns
  For a full list of skipped files, run semgrep with the --verbose flag.

(need more rules? `semgrep login` for additional free Semgrep Registry rules)

Ran 1 rule on 1220 files: 0 findings.
If Semgrep missed a finding, please send us feedback to let us know!
  $ semgrep shouldafound --help

```

**Comment by 06kellyjac** [comment]
Thanks for testing on both platforms :heart:

File: pkgs/tools/security/semgrep/update.sh [hunk 2]
```diff
@@ -50,43 +50,54 @@ fetchgithub() {
...
+  nix-prefetch "{ fetchPypi }:
+fetchPypi rec {
+  pname = \"semgrep\";
+  version = \"$VERSION\";
+  format = \"wheel\";
+  dist = python;
+  python = \"cp37.cp38.cp39.py37.py38.py39\";
+  platform = \"$PLATFORM\";
+}
+"
```
Comment (ambroisie) [inline pkgs/tools/security/semgrep/update.sh:65]: I believe you need to add it to the `nix-shell` list of packages?

**Review by ambroisie — APPROVED** [review]
Not sure about `nix-prefetch`, otherwise LGTM.

File: pkgs/tools/security/semgrep/semgrep-core.nix [hunk 3]
```diff
@@ -1,21 +1,48 @@
...
+
+  nativeBuildInputs = [
+    unzip
+  ];
+
+  unpackPhase = ''
+    runHook preUnpack
+    unzip $src
+    runHook postUnpack
+  '';
```
Comment (SuperSandro2000) [inline pkgs/tools/security/semgrep/semgrep-core.nix:35]: Isn't this the default when unzip is added to nativeBuildInputs?

File: pkgs/tools/security/semgrep/default.nix [hunk 4]
```diff
@@ -97,10 +104,12 @@ buildPythonApplication rec {
 
   passthru = {
     inherit common;
+    submodulesSubset = lib.mapAttrs (k: args: fetchFromGitHub args) common.submodules;
     updateScript = ./update.sh;
   };
 
   meta = common.meta // {
     description = common.meta.description + " - cli";
+    platforms = semgrep-core.meta.platforms;
```
Comment (SuperSandro2000) [inline pkgs/tools/security/semgrep/default.nix:113]:
```suggestion
    inherit (semgrep-core.meta) platforms;
```

File: pkgs/tools/security/semgrep/semgrep-core.nix (hunk 3 above)
Comment (06kellyjac) [reply pkgs/tools/security/semgrep/semgrep-core.nix:35]: Not sure. I thought it got upset because the extension is .whl

I can try again without

Comment (06kellyjac) [reply pkgs/tools/security/semgrep/semgrep-core.nix]: Tried it and yeah the unzip unpacker checks for a `.zip` extension.

I guess we could modify it to accept `.whl` and maybe `.ZIP` while we're at it

https://github.com/NixOS/nixpkgs/blob/master/pkgs/tools/archivers/unzip/setup-hook.sh#L3

Comment (06kellyjac) [reply pkgs/tools/security/semgrep/semgrep-core.nix]: For now I have added a comment

File: pkgs/tools/security/semgrep/update.sh (hunk 2 above)
Comment (06kellyjac) [reply pkgs/tools/security/semgrep/update.sh:65]: you're right. added

**Comment by emilytrau** [comment]
Resolves? #188944

**Comment by 06kellyjac** [comment]
yes. I wasn't aware of that issue

File: pkgs/tools/security/semgrep/semgrep-core2.nix [hunk 5]
```diff
@@ -0,0 +1,73 @@
+{ lib, stdenv, fetchFromGitHub, dune_3, pcre, ocamlPackages }:
```
Comment (SuperSandro2000) [inline pkgs/tools/security/semgrep/semgrep-core2.nix:1]: Did you commit this file by accident?

Comment (06kellyjac) [reply pkgs/tools/security/semgrep/semgrep-core2.nix:1]: Yes lol. Ill remove it when im at my pc

Comment (06kellyjac) [reply pkgs/tools/security/semgrep/semgrep-core2.nix]: removed

**Review by ambroisie — APPROVED** [review]

**Comment by DamienCassou** [comment]
Anything preventing the merge of this PR?

**Comment by DamienCassou** [comment]
Result of `nixpkgs-review pr 235466` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>semgrep</li>
    <li>semgrep-core</li>
    <li>semgrep.dist</li>
  </ul>
</details>

**Review by Janik-Haag — APPROVED** [review]
Build's and runs fine locally. One maintainer already approved this so I will go ahead and merge this.

**Comment by Janik-Haag** [comment]
@06kellyjac why did you add the backport label?

**Comment by 06kellyjac** [comment]
Why not?

**Comment by Janik-Haag** [comment]
Because we (at least as far as I know) only backport secrity and bug fixes but don't back port version updates.

**Comment by DamienCassou** [comment]
Here is what the manual as to say about backports: https://nixos.org/manual/nixpkgs/stable/#acceptable-backport-criteria.

**Comment by 06kellyjac** [comment]
Thanks for linking @DamienCassou 

So yes, this falls under

> Version updates which include new functionality (but no breaking changes) 

Most people wont add the backport label because its extra effort, but in this case semgrep was stuck on 1.15.0 for ages and there was a version update request issue.
If it isnt backported 23.05 will have the same old version of semgrep as 22.11 all the way until 23.11

**Comment by Janik-Haag** [comment]
Okay thanks for explaining, can you test and approve the backport?

**Pull request merged**